akka-http-aggregator/src/main/scala/com/pagerduty/akka/http/aggregator/aggregator/GenericAggregator.scala

```diff
@@ -0,0 +1,133 @@
+import com.pagerduty.akka.http.aggregator.AggregatorUpstream
+import com.pagerduty.akka.http.headerauthentication.model.HeaderAuthConfig
+import com.pagerduty.akka.http.proxy.HttpProxy
+
+import scala.concurrent.{ExecutionContext, Future}
+
+trait GenericAggregator[RequestKey, AccumulatedState, AddressingConfig]
+    extends Aggregator[AddressingConfig] {
+  type RequestMap =
+    Map[RequestKey, (AggregatorUpstream[AddressingConfig], HttpRequest)]
+  type ResponseMap = Map[RequestKey, (HttpResponse, String)]
+  type AccumulatedStateAndRequests = (AccumulatedState, RequestMap)
+  type HandlerResult = Either[HttpResponse, AccumulatedStateAndRequests]
+  type ResponseHandler = (AccumulatedState, ResponseMap) => HandlerResult
+
+  // implement these methods
+  def handleIncomingRequest(
+      authConfig: HeaderAuthConfig
+  )(incomingRequest: HttpRequest, authData: authConfig.AuthData): HandlerResult
+
+  def intermediateResponseHandlers: Seq[ResponseHandler]
+
+  def buildOutgoingResponse(accumulatedState: AccumulatedState,
+                            upstreamResponses: ResponseMap): HttpResponse
+
+  // implementation
+  def execute(authConfig: HeaderAuthConfig)(authedRequest: HttpRequest,
+                                            authData: authConfig.AuthData)(
+      implicit httpProxy: HttpProxy[AddressingConfig],
+      executionContext: ExecutionContext,
+      materializer: Materializer): Future[HttpResponse] = {
+    val initialHandlerResult =
+      handleIncomingRequest(authConfig)(authedRequest, authData)
+
+    initialHandlerResult match {
+      case Right((initialState, initialRequests)) =>
+        // the initial handler returned some requests, execute them
+        val fInitialResponses = executeRequests(authConfig)(initialState,
+                                                            initialRequests,
+                                                            authedRequest)
+
+        val intermediateHandlersResult = executeIntermediateHandlers(
+          authConfig)(authedRequest, fInitialResponses)
+
+        // all the intermediate handlers have run (or short-circuited), tell the aggregator to build the outgoing response
+        intermediateHandlersResult.map {
+          case Right((finalState, responseMap)) =>
+            // we don't have a response yet, have the aggregator do the final response build
+            buildOutgoingResponse(finalState, responseMap)
+          case Left(shortCircuit) =>
+            // we already have the final (short-circuit) response
+            shortCircuit
+        }
+      case Left(httpResponse) =>
+        // the initial handler short-circuited and provided a response from the initial request, don't run intermediate handlers or final builder
+        Future.successful(httpResponse)
+    }
+  }
+
+  private def executeIntermediateHandlers(authConfig: HeaderAuthConfig)(
+      authedRequest: HttpRequest,
+      fInitialStateAndResponses: Future[(AccumulatedState, ResponseMap)])(
```

Comment:
Are we using Hungarian notation for `Future`s now? 😉

Reply:
Hah! You know at one point over this code's evolution I think there was a reason.... probably I wanted to have `fThing` and `thing` be distinct values, but really there's no reason for it now. Gonna drop `em.